crates/re_viewer/src/ui/view_spatial/scene/scene_part/images.rs

```diff
@@ -89,63 +87,69 @@ fn to_textured_rect(
+        self.draw_order == draw_order && Self::is_plane_similar(self.plane, plane)
+    }
+}
+
 fn handle_image_layering(scene: &mut SceneSpatial) {
     crate::profile_function!();
 
     // Handle layered rectangles that are on (roughly) the same plane with the same depth offset and were logged in sequence.
-    // First, group by similar plane.
-    // TODO(andreas): Need planes later for picking as well!
-    let images_grouped_by_plane = {
-        let mut cur_plane = macaw::Plane3::from_normal_dist(Vec3::NAN, std::f32::NAN);
-        let mut rectangle_group: Vec<Image> = Vec::new();
-        scene
-            .primitives
-            .images
-            .drain(..) // We rebuild the list as we might reorder as well!
-            .batching(move |it| {
-                for image in it {
-                    let rect = &image.textured_rect;
-
-                    let prev_plane = cur_plane;
-                    cur_plane = macaw::Plane3::from_normal_point(
-                        rect.extent_u.cross(rect.extent_v).normalize(),
-                        rect.top_left_corner_position,
-                    );
-
-                    fn is_plane_similar(a: macaw::Plane3, b: macaw::Plane3) -> bool {
-                        a.normal.dot(b.normal) > 0.99 && (a.d - b.d).abs() < 0.01
-                    }
-
-                    // Use draw order, not depth offset since depth offset might change when draw order does not.
-                    let has_same_draw_order = rectangle_group
-                        .last()
-                        .map_or(true, |last_image| last_image.draw_order == image.draw_order);
-
-                    // If the planes are similar, add them to the same group, otherwise start a new group.
-                    if has_same_draw_order && is_plane_similar(prev_plane, cur_plane) {
-                        rectangle_group.push(image);
-                    } else {
-                        let previous_group = std::mem::replace(&mut rectangle_group, vec![image]);
-                        return Some(previous_group);
-                    }
-                }
-                if !rectangle_group.is_empty() {
-                    Some(rectangle_group.drain(..).collect())
-                } else {
-                    None
-                }
-            })
+    // Rectangles that have different draw order are not grouped together.
+    let mut image_groups: Vec<ImageGroup> = Vec::new();
+
+    // We rebuild the list as we might reorder as well!
+    for image in scene.primitives.images.drain(..) {
+        let rect = &image.textured_rect;
+        let plane = macaw::Plane3::from_normal_point(
+            rect.extent_u.cross(rect.extent_v).normalize(),
+            rect.top_left_corner_position,
+        );
+
+        if let Some(group) = image_groups
+            .iter_mut()
+            .find(|group| group.is_part_of_group(image.draw_order, plane))
```

Comment:
Do we need really need the ugly `is_plane_similar`? Isn't equality of the `Plane3` better? Then we can put them in a `BTreeMap` to avoid the O(N²) behavior here